app/models/attachment_reference.rb

```diff
@@ -0,0 +1,38 @@
+class AttachmentReference < ActiveRecord::Base
+  belongs_to :attachable, polymorphic: true, inverse_of: :attachment_references
+  belongs_to :attachment, inverse_of: :attachment_references
+
+  # Get the name from the file and then further build or find an attachment based on file's SHA256
+  # hash.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  def file=(file)
+    self.name = filename(file)
+    self.attachment = find_or_initialize_attachment_by(file)
+  end
+
+  private
+
+  # Calculates the SHA256 hash and return the existing attachment which has the same hash.
+  # Otherwise a new attachment will be built.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  # @return [Attachment] The attachment which contains the file.
+  def find_or_initialize_attachment_by(file)
```

Comment:
We use Class.find_or_initialize_by() right? Should there be another separate method?